config/auth_local_metadata_service.go

```diff
@@ -0,0 +1,171 @@
+//	access_token: The requested access token.
+//	token_type: The type of token, which is a "Bearer" access token.
+//	expires_on: The timespan when the access token expires.
+type LocalMetadataServiceCredentials struct{}
+
+func (c LocalMetadataServiceCredentials) Name() string {
+	return "local-metadata-service"
+}
+
+func (c LocalMetadataServiceCredentials) Configure(ctx context.Context, cfg *Config) (func(*http.Request) error, error) {
+	if cfg.LocalMetadataServiceUrl == "" {
+		return nil, nil
+	}
+
+	parsedMetadataServiceUrl, err := url.Parse(cfg.LocalMetadataServiceUrl)
+	if err != nil {
+		return nil, fmt.Errorf("invalid auth server URL: %w", err)
+	}
+
+	// only allow localhost URLs
+	if parsedMetadataServiceUrl.Hostname() != "localhost" && parsedMetadataServiceUrl.Hostname() != "127.0.0.1" {
+		return nil, fmt.Errorf("invalid auth server URL: %s", cfg.LocalMetadataServiceUrl)
+	}
+
+	resp, err := makeRequest(parsedMetadataServiceUrl)
+	if err != nil {
+		return nil, nil
+	}
+
+	if resp == nil {
+		return nil, nil
+	}
+
+	if resp.Host == "" {
+		return nil, nil
+	}
+
+	if cfg.Host != "" && cfg.Host != resp.Host {
+		return nil, nil
+	}
+
+	cfg.Host = resp.Host
+
+	ts := metadataServiceTokenSource{
+		metadataServiceURL: parsedMetadataServiceUrl,
+		databricksHost:     cfg.Host,
+	}
+	return refreshableVisitor(&ts), nil
+}
+
+// makes a request to the server and returns the token
+func makeRequest(serverUrl *url.URL) (*serverResponse, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), metadataServiceTimeout)
+	defer cancel()
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, serverUrl.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("token request: %w", err)
+	}
+	req.Header.Add("Metadata", "true")
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("token response: %w", err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode == 404 {
+		return nil, nil
+	}
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("token read: %w", err)
+	}
+	if resp.StatusCode != 200 {
+		return nil, fmt.Errorf("token error: %s", body)
+	}
+
+	var response serverResponse
+	if err := json.Unmarshal(body, &response); err != nil {
+		return nil, fmt.Errorf("token parse: %w", err)
+	}
+
+	// parse and normalize url
+	parsedHost, err := url.Parse(response.Host)
+	if err != nil {
+		return nil, fmt.Errorf("token parse: Invalid URL %s", response.Host)
+	}
+
+	response.Host = parsedHost.String()
+
+	return &response, nil
+}
+
+type metadataServiceTokenSource struct {
+	metadataServiceURL *url.URL
+	databricksHost     string
+}
+
+func (t metadataServiceTokenSource) Token() (*oauth2.Token, error) {
+	token, err := makeRequest(t.metadataServiceURL)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if token.Host != t.databricksHost {
+		return nil, fmt.Errorf("host in token (%s) doesn't match configured host (%s)", token.Host, t.databricksHost)
+	}
```

Comment:
We should take it one step further and add an opaque "session" identifier or something that we can check.

It is possible that the session changes (e.g. other metadata we add later) while the host remains the same.

If that is the case this session should also be invalidated.